### PR TITLE
Refactor downloader package dist path parsing

### PR DESCRIPTION
**Pull request by sshymko**
- Remove unused argument from method `FileDownloader::getFileName()`
- Reuse package dist path parsing of filename and extension

### CODE REVIEW

File: src/Composer/Downloader/FileDownloader.php [hunk 1]
```diff
@@ -368,6 +368,12 @@ public function install(PackageInterface $package, string $path, bool $output =
         return \React\Promise\resolve(null);
     }
 
+    protected function getDistPath(PackageInterface $package, int $component): string
+    {
+        // @phpstan-ignore-next-line
```
Comment (sshymko) [inline src/Composer/Downloader/FileDownloader.php:373]: Suppress the following false-positive error by PHPStan:
> Method Composer\Downloader\FileDownloader::getDistPath() should return string but returns array<string, string>|string.

**Comment by Seldaek** [comment]
What's the point of this refactoring? It's changing the API  of downloaders, which are likely extended somewhere else, possibly breaking things.. so I need a good reason.

**Comment by sshymko** [comment]
@Seldaek 
No grand idea behind the refactoring :) Was browsing the codebase and noticed an unused argument. You're right that it breaks backwards compatibility. Wasn't sure what Composer's stance on BC is in terms of inheritance. Composer is typically used as a tool (executable phar) rather than as a regular PHP library. Would this type of change be more suitable for the next major version release?

While removing the unused argument identified duplication of the dist path parsing logic, so implemented its reuse. Would you be interested in keeping this part of the refactoring? Can revert the unused argument. The new method can be private, it's been marked protected solely for consistency with the nearby code.

**Comment by stof** [comment]
@sshymko Composer allows plugins. So even when used as an executable, its PHP API can still be used by those plugins..

**Comment by sshymko** [comment]
Okay, reverted removal of the unused argument in d52377a5af4c8df6b76e2320eb6d4c02c3cae372.

Keeping the PR with the remaining (admittedly small) changes, since it's already been implemented.

File: src/Composer/Downloader/FileDownloader.php [hunk 2]
```diff
@@ -431,7 +437,7 @@ public function remove(PackageInterface $package, string $path, bool $output = t
      */
     protected function getFileName(PackageInterface $package, string $path): string
     {
-        return rtrim($this->config->get('vendor-dir').'/composer/tmp-'.md5($package.spl_object_hash($package)).'.'.pathinfo(parse_url(strtr((string) $package->getDistUrl(), '\\', '/'), PHP_URL_PATH), PATHINFO_EXTENSION), '.');
+        return rtrim($this->config->get('vendor-dir') . '/composer/tmp-' . md5($package . spl_object_hash($package)) . '.' . $this->getDistPath($package, PATHINFO_EXTENSION), '.');
```
[the remark this answers is not in the document]
Comment (Seldaek) [reply src/Composer/Downloader/FileDownloader.php:440]: That's very much a habit/personal taste thing IMO. I for one am not used to sprintf and it tends to bug me when I read those..

[the remark this answers is not in the document]
Comment (drupol) [reply src/Composer/Downloader/FileDownloader.php:440]: I think you're right, I'm using `sprintf` all the time, maybe that's the reason why seeing this confuses me.

**Comment by Seldaek** [comment]
Thanks

**Pull request merged**